cai_causal_graph/causal_graph.py

```diff
@@ -2148,12 +2149,21 @@ def from_adjacency_matrix(
         graph = cls()
         graph.add_nodes_from(nodes)
         for i, j in itertools.combinations(range(len(nodes)), 2):
+            # Keep validate as False to not slow construction. We will look at validate param after graph is constructed.
             if adjacency[i, j] != 0 and adjacency[j, i] == 0:
-                graph.add_edge(nodes[i], nodes[j], edge_type=EdgeType.DIRECTED_EDGE)
+                graph.add_edge(nodes[i], nodes[j], edge_type=EdgeType.DIRECTED_EDGE, validate=False)
             elif adjacency[i, j] == 0 and adjacency[j, i] != 0:
-                graph.add_edge(nodes[j], nodes[i], edge_type=EdgeType.DIRECTED_EDGE)
+                graph.add_edge(nodes[j], nodes[i], edge_type=EdgeType.DIRECTED_EDGE, validate=False)
             elif adjacency[i, j] != 0 and adjacency[j, i] != 0:
-                graph.add_edge(nodes[i], nodes[j], edge_type=EdgeType.UNDIRECTED_EDGE)
+                graph.add_edge(nodes[i], nodes[j], edge_type=EdgeType.UNDIRECTED_EDGE, validate=False)
+
+        # check that there are no cycles of directed edges
+        if validate:
```

Comment:
In general, I think we decided at some point that a "causal graph" should not have cycles which is why we don't allow the object construction. `networkx` is just generic graph so it is fine but because we are building "causal graphs" we did not want to support it. Probably ok to revisit that decision though as I don't think we need to be so strict.